File: 2020/day10.py
from typing import Callable
from os.path import isfile, join as path_join
file_name = path_join('input', 'day10.txt')
def to_list(mf: Callable = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: Callable = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])

data = to_list()

data = sorted(data)

def part1():
	a = [None, 0, 0, 0]

	a[data[0]] += 1

	for i in range(len(data) - 1):
		a[data[i+1] - data[i]] += 1

	a[3] += 1

	print(a[1] * a[3])


part1()

data.append(data[-1] + 3)

from collections import defaultdict

def part2():
	ways = defaultdict(int)
	ways[0] = 1
	for v in data:
		ways[v] = ways[v-1] + ways[v-2] + ways[v-3]

	print(ways[data[-1]])


part2()

File: cpp/2020/day10.cpp
#include "../aocHelper.h"

class Day10 : public BaseDay
{
public:
	Day10() : BaseDay("10") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		vector<int> data;

		int n;

		while ((n = numericParse<int>(input)))
		{
			data.push_back(n);
		}

		sort(data.begin(), data.end());

		// part 1

		int a[4] = { 0, 0, 0, 0 };

		a[data[0]]++;

		for (int i = 0; i < data.size() - 1; i++)
		{
			a[data[i + 1] - data[i]]++;
		}

		a[3]++;

		part1 = a[1] * a[3];

		// part 2

		data.push_back(data.back() + 3);

		long long ways[200] = { 1 };

		for (auto& v : data)
		{
			long long s = 0;
			if (v > 0)
			{
				s += ways[v - 1];
			}
			if (v > 1)
			{
				s += ways[v - 2];
			}
			if (v > 2)
			{
				s += ways[v - 3];
			}

			ways[v] = s;
		}

		part2 = ways[data.back()];

		return { part1, part2 };
	}
};
